src/test/java/com/google/devtools/build/bfg/JavaSourceFileParserTest.java

```diff
@@ -274,7 +288,47 @@ public void tolerateEmptyFiles() throws Exception {
     Path file = writeFile(workspace.resolve("com/hello/package-info.java"), "package com.hello;");
     JavaSourceFileParser parser = createParser(file);
 
-    assertThatGraphsEqual(parser.getClassDependencyGraph(), newGraph());
+    assertThatGraphsEqual(parser.getClassToClass(), newGraph());
+  }
+
+  @Test
+  public void ruleKindDetection() throws Exception {
```

Comment:
nit: can you put in separate file and separate into smaller tests?

I think 
```java
abstractClassDetectionTest fails
```
to be more informative than 
```java
ruleKindDetection fails at line X
``` 
with line x being some assertion.

as described above, by making ruleKindDetection into to a separate class, each of these can be a separate unit test without clogging up this test file. This I believe is better once we add support for more languages.

Reply:
I think the test file is still reasonably small. If there's more functionality put into `decideRuleKind` we can reconsider.
Also, other languages won't use this parser.

Reply:
> Also, other languages won't use this parser.

Oh. Then yeah, dismiss previous suggestion.